conflictlog/pool.go

```diff
@@ -2,82 +2,248 @@ package conflictlog
+// Get returns an object from the pool. If there is none then it creates
+// one by calling newConnFn() and returns it.
+func (pool *connPool) Get(bucketName string) (conn io.Closer, err error) {
+	conn = pool.get(bucketName)
+	if conn != nil {
 		return
 	}
 
+	conn, err = pool.newConnFn(bucketName)
 	return
 }
 
-func (pool *writerPool) release(w Writer) {
-	wlist := pool.getList(w.Bucket())
-	wlist.push(w)
+func (pool *connPool) Put(bucketName string, conn io.Closer) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	l := pool.getOrCreateListNoLock(bucketName)
+	l.push(conn)
+}
+
+func (pool *connPool) Close() {
+
+}
+
+// reapConnList collects all the connection lists which have not been used for the reapInterval time
+// This does not close the connection itself
+func (pool *connPool) reapConnList() []*connList {
+	connListList := []*connList{}
+	reapedBuckets := []string{}
+
+	now := time.Now()
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	// Collect all expired buckets and its lists
+	for bucketName, connList := range pool.buckets {
+		elapsed := now.Sub(connList.lastUsed)
+		if elapsed >= pool.reapInterval {
+			reapedBuckets = append(reapedBuckets, bucketName)
+			connListList = append(connListList, connList)
+		}
+	}
+
+	// Delete expired buckets
+	// It is possible that a connection for a deleted bucket will be requested
+	// after this. This is fine since it will treated like a new bucket being requested
+	// for the first time
+	for _, bucketName := range reapedBuckets {
+		pool.logger.Debugf("reaping connections for bucket=%s", bucketName)
+		delete(pool.buckets, bucketName)
+	}
+
+	return connListList
+}
+
+// gcOnce runs one single iteration of reaping the connections
+func (pool *connPool) gcOnce() {
+	connListList := pool.reapConnList()
+
+	// Note: the closing of the connections happen outside the pool lock.
+	// From this point, a parallel request to create a connection is safe
+	// and it will land in a new connList.
+	for _, connList := range connListList {
+		connList.closeAll()
+	}
+}
+
+// gc reaps the unused connections by checking at a regular interval
+// This is how it works
+//  1. Every time connList is used (pop or push) it updates its lastUsed = time.Now()
+//  2. Each iteration of GC checks all buckets and its list to see which ones have exceeded the reapInterval
+//  3. It collects such lists and for each of the lists it closes all the connections
+//
+// It is certainly possible to get a access pattern such that a connList is being closed and
+// there is a parallel request connection for the same bucket. This request can be catered to in parallel safely
+// as it will land in a completely new list. This cannot be avoided and it is expected that
+// such cases will be much fewer.
+func (pool *connPool) gc() {
+	for {
+		// we copy the ticker so that it can be modified in parallel
+		gcTicker := pool.getGCTicker()
+
+		select {
+		case <-pool.finch:
+			pool.logger.Info("conflict log conn pool gc worker exiting")
```

Comment:
Should there be an ID for this connPool and print it as part of start/stop for message matching?